clippy_lints/src/explicit_into_iter_fn_arg.rs

```diff
@@ -0,0 +1,144 @@
+    ///
+    /// ### Example
+    /// ```rust
+    /// fn even_sum<I: IntoIterator<Item = u32>>(iter: I) -> u32 {
+    ///     iter.into_iter().filter(|&x| x % 2 == 0).sum()
+    /// }
+    ///
+    /// let _ = even_sum(vec![1, 2, 3].into_iter());
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn even_sum<I: IntoIterator<Item = u32>>(iter: I) -> u32 {
+    ///     iter.into_iter().filter(|&x| x % 2 == 0).sum()
+    /// }
+    ///
+    /// let _ = even_sum(vec![1, 2, 3]);
+    /// ```
+    #[clippy::version = "1.71.0"]
+    pub EXPLICIT_INTO_ITER_FN_ARG,
+    pedantic,
+    "explicit call to `.into_iter()` in function argument accepting `IntoIterator`"
+}
+declare_lint_pass!(ExplicitIntoIterFnArg => [EXPLICIT_INTO_ITER_FN_ARG]);
+
+enum MethodOrFunction {
+    Method,
+    Function,
+}
+
+/// Returns the span of the `IntoIterator` trait bound in the function pointed to by `fn_did`
+fn into_iter_bound(cx: &LateContext<'_>, fn_did: DefId, param_index: u32) -> Option<Span> {
+    if let Some(into_iter_did) = cx.tcx.get_diagnostic_item(sym::IntoIterator) {
+        cx.tcx
+            .predicates_of(fn_did)
+            .predicates
+            .iter()
+            .find_map(|&(ref pred, span)| {
+                if let ty::PredicateKind::Clause(ty::Clause::Trait(tr)) = pred.kind().skip_binder()
+                    && tr.def_id() == into_iter_did
+                    && tr.self_ty().is_param(param_index)
+                {
+                    Some(span)
+                } else {
+                    None
+                }
+            })
+    } else {
+        None
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for ExplicitIntoIterFnArg {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if expr.span.from_expansion() {
+            return;
+        }
+
+        if let ExprKind::MethodCall(name, recv, ..) = expr.kind
+            && is_trait_method(cx, expr, sym::IntoIterator)
+            && name.ident.name == sym::into_iter
+            && let Some(parent_expr) = get_parent_expr(cx, expr)
+        {
+            let parent_expr = match parent_expr.kind {
+                ExprKind::Call(recv, args) if let ExprKind::Path(ref qpath) = recv.kind => {
+                    cx.qpath_res(qpath, recv.hir_id).opt_def_id()
+                        .map(|did| (did, args, MethodOrFunction::Function))
+                }
+                ExprKind::MethodCall(.., args, _) => {
+                    cx.typeck_results().type_dependent_def_id(parent_expr.hir_id)
+                        .map(|did| (did, args, MethodOrFunction::Method))
+                }
+                _ => None,
+            };
+
+            if let Some((parent_fn_did, args, kind)) = parent_expr
+                && let sig = cx.tcx.fn_sig(parent_fn_did).skip_binder().skip_binder()
+                && let Some(arg_pos) = args.iter().position(|x| x.hir_id == expr.hir_id)
+                && let Some(&into_iter_param) = sig.inputs().get(match kind {
```

Comment:
You could impl `Add<usize>` for `MethodOrFunction` to simplify this, but it won't block the PR from merging.

Reply:
Hmm, I feel like an `Add` impl could potentially be a little confusing? I personally wouldn't know what it meant to add `MethodOrFunction` to a usize. Maybe it could just be a normal method? i.e.
```rs
impl MethodOrFunction {
    /// Maps the argument positon in `pos` to the parameter position.
    /// For methods, the `self` argument is skipped.
    fn arg_pos(&self, pos: usize) -> usize {
        match self {
            MethodOrFunction::Method => pos + 1,
            MethodOrFunction::Function => pos,
        }
    }
}
```